File: bindings/lua/satox_bindings/thread_safe_manager.lua
-- Thread-Safe Manager for Lua Bindings
-- Provides comprehensive thread safety using Lua's concurrency primitives

local ffi = require("ffi")
local bit = require("bit")

-- Load threading library (LuaJIT or compatible)
local thread = require("thread") or require("lanes") or require("copas")

-- Thread state enumeration
local ThreadState = {
    UNINITIALIZED = "uninitialized",
    INITIALIZING = "initializing", 
    INITIALIZED = "initialized",
    SHUTTING_DOWN = "shutting_down",
    SHUTDOWN = "shutdown",
    ERROR = "error"
}

-- Thread safety statistics
local ThreadSafeStats = {}
ThreadSafeStats.__index = ThreadSafeStats

function ThreadSafeStats.new()
    local self = setmetatable({}, ThreadSafeStats)
    self.total_operations = 0
    self.concurrent_operations = 0
    self.max_concurrent_operations = 0
    self.average_operation_time = 0.0
    self.total_operation_time = 0.0
    self.lock_contention_count = 0
    self.last_operation_time = 0.0
    return self
end

-- Thread-safe manager base class
local ThreadSafeManager = {}
ThreadSafeManager.__index = ThreadSafeManager

function ThreadSafeManager.new(name)
    name = name or "ThreadSafeManager"
    local self = setmetatable({}, ThreadSafeManager)
    
    self.name = name
    self.state = ThreadState.UNINITIALIZED
    self.initialized = false
    self.disposed = false
    self.last_error = ""
    self.error_count = 0
    
    -- Thread safety primitives
    self.state_lock = thread.newlock()
    self.error_lock = thread.newlock()
    self.stats_lock = thread.newlock()
    self.operation_lock = thread.newlock()
    self.event_lock = thread.newlock()
    self.resource_lock = thread.newlock()
    self.config_lock = thread.newlock()
    self.cache_lock = thread.newlock()
    
    -- Statistics
    self.stats = ThreadSafeStats.new()
    self.operation_start_times = {}
    self.active_operations = 0
    self.operation_counter = 0
    
    -- Event handling
    self.event_handlers = {}
    self.event_stats = {}
    
    -- Resource management
    self.resources = {}
    
    -- Configuration
    self.config = {}
    
    -- Cache
    self.cache = {}
    self.cache_timestamps = {}
    
    -- Performance monitoring
    self.performance_enabled = true
    self.start_time = os.time()
    
    return self
end

function ThreadSafeManager:ensure_initialized()
    self.state_lock:lock()
    if self.state == ThreadState.UNINITIALIZED then
        self:set_error("Manager not initialized")
        self.state_lock:unlock()
        return false
    elseif self.state == ThreadState.SHUTDOWN then
        self:set_error("Manager has been disposed")
        self.state_lock:unlock()
        return false
    elseif self.state == ThreadState.ERROR then
        self:set_error("Manager is in error state")
        self.state_lock:unlock()
        return false
    end
    self.state_lock:unlock()
    return true
end

function ThreadSafeManager:ensure_not_disposed()
    self.state_lock:lock()
    if self.state == ThreadState.SHUTDOWN then
        self:set_error("Manager has been disposed")
        self.state_lock:unlock()
        return false
    end
    self.state_lock:unlock()
    return true
end

function ThreadSafeManager:set_state(state)
    self.state_lock:lock()
    self.state = state
    self.state_lock:unlock()
end

function ThreadSafeManager:get_state()
    self.state_lock:lock()
    local state = self.state
    self.state_lock:unlock()
    return state
end

function ThreadSafeManager:wait_for_state(target_state, timeout)
    timeout = timeout or 5.0
    local start_time = os.time()
    
    while os.time() - start_time < timeout do
        if self:get_state() == target_state then
            return true
        end
        thread.sleep(0.01)
    end
    return false
end

function ThreadSafeManager:begin_operation()
    if not self.performance_enabled then
        return 0
    end
    
    self.stats_lock:lock()
    local operation_id = self.operation_counter
    self.operation_counter = self.operation_counter + 1
    self.active_operations = self.active_operations + 1
    self.operation_start_times[operation_id] = os.time()
    self.stats.concurrent_operations = self.active_operations
    self.stats.max_concurrent_operations = math.max(self.stats.max_concurrent_operations, self.active_operations)
    self.stats_lock:unlock()
    
    return operation_id
end

function ThreadSafeManager:end_operation(operation_id)
    if not self.performance_enabled or operation_id == 0 then
        return
    end
    
    self.stats_lock:lock()
    if self.operation_start_times[operation_id] then
        local operation_time = os.time() - self.operation_start_times[operation_id]
        self.operation_start_times[operation_id] = nil
        
        self.active_operations = self.active_operations - 1
        self.stats.total_operations = self.stats.total_operations + 1
        self.stats.total_operation_time = self.stats.total_operation_time + operation_time
        self.stats.average_operation_time = self.stats.total_operation_time / self.stats.total_operations
        self.stats.last_operation_time = os.time() - self.start_time
        self.stats.concurrent_operations = self.active_operations
    end
    self.stats_lock:unlock()
end

function ThreadSafeManager:set_error(error)
    self.error_lock:lock()
    self.last_error = error
    self.error_count = self.error_count + 1
    self:set_state(ThreadState.ERROR)
    self.error_lock:unlock()
end

function ThreadSafeManager:clear_error()
    self.error_lock:lock()
    self.last_error = ""
    if self.state == ThreadState.ERROR then
        self:set_state(ThreadState.INITIALIZED)
    end
    self.error_lock:unlock()
end

function ThreadSafeManager:get_last_error()
    self.error_lock:lock()
    local error = self.last_error
    self.error_lock:unlock()
    return error
end

function ThreadSafeManager:get_error_count()
    self.error_lock:lock()
    local count = self.error_count
    self.error_lock:unlock()
    return count
end

function ThreadSafeManager:is_initialized()
    return self:get_state() == ThreadState.INITIALIZED
end

function ThreadSafeManager:is_disposed()
    return self:get_state() == ThreadState.SHUTDOWN
end

function ThreadSafeManager:get_stats()
    self.stats_lock:lock()
    local stats = {
        total_operations = self.stats.total_operations,
        concurrent_operations = self.stats.concurrent_operations,
        max_concurrent_operations = self.stats.max_concurrent_operations,
        average_operation_time = self.stats.average_operation_time,
        total_operation_time = self.stats.total_operation_time,
        lock_contention_count = self.stats.lock_contention_count,
        last_operation_time = self.stats.last_operation_time
    }
    self.stats_lock:unlock()
    return stats
end

function ThreadSafeManager:reset_stats()
    self.stats_lock:lock()
    self.stats = ThreadSafeStats.new()
    self.operation_start_times = {}
    self.active_operations = 0
    self.operation_counter = 0
    self.stats_lock:unlock()
end

function ThreadSafeManager:enable_performance_monitoring()
    self.performance_enabled = true
end

function ThreadSafeManager:disable_performance_monitoring()
    self.performance_enabled = false
end

function ThreadSafeManager:get_uptime()
    return os.time() - self.start_time
end

function ThreadSafeManager:set_config(key, value)
    self.config_lock:lock()
    self.config[key] = value
    self.config_lock:unlock()
end

function ThreadSafeManager:get_config(key, default)
    self.config_lock:lock()
    local value = self.config[key] or default
    self.config_lock:unlock()
    return value
end

function ThreadSafeManager:set_cache(key, value, ttl)
    ttl = ttl or 300.0
    self.cache_lock:lock()
    self.cache[key] = value
    self.cache_timestamps[key] = os.time() + ttl
    self.cache_lock:unlock()
end

function ThreadSafeManager:get_cache(key, default)
    self.cache_lock:lock()
    if self.cache[key] then
        if os.time() < self.cache_timestamps[key] then
            local value = self.cache[key]
            self.cache_lock:unlock()
            return value
        else
            -- Expired, remove from cache
            self.cache[key] = nil
            self.cache_timestamps[key] = nil
        end
    end
    self.cache_lock:unlock()
    return default
end

function ThreadSafeManager:clear_cache()
    self.cache_lock:lock()
    self.cache = {}
    self.cache_timestamps = {}
    self.cache_lock:unlock()
end

function ThreadSafeManager:add_event_handler(event, handler)
    self.event_lock:lock()
    if not self.event_handlers[event] then
        self.event_handlers[event] = {}
    end
    table.insert(self.event_handlers[event], handler)
    self.event_lock:unlock()
end

function ThreadSafeManager:remove_event_handler(event, handler_index)
    self.event_lock:lock()
    if self.event_handlers[event] and self.event_handlers[event][handler_index] then
        table.remove(self.event_handlers[event], handler_index)
    end
    self.event_lock:unlock()
end

function ThreadSafeManager:emit_event(event, ...)
    local handlers = {}
    self.event_lock:lock()
    if self.event_handlers[event] then
        handlers = self.event_handlers[event]
        self.event_stats[event] = (self.event_stats[event] or 0) + 1
    end
    self.event_lock:unlock()
    
    -- Call handlers outside of lock to prevent deadlocks
    for _, handler in ipairs(handlers) do
        local ok, err = pcall(handler, ...)
        if not ok then
            self:set_error("Error in event handler: " .. tostring(err))
        end
    end
end

function ThreadSafeManager:add_resource(name, resource)
    self.resource_lock:lock()
    self.resources[name] = resource
    self.resource_lock:unlock()
end

function ThreadSafeManager:get_resource(name)
    self.resource_lock:lock()
    local resource = self.resources[name]
    self.resource_lock:unlock()
    return resource
end

function ThreadSafeManager:remove_resource(name)
    self.resource_lock:lock()
    self.resources[name] = nil
    self.resource_lock:unlock()
end

function ThreadSafeManager:clear_resources()
    self.resource_lock:lock()
    self.resources = {}
    self.resource_lock:unlock()
end

function ThreadSafeManager:dispose()
    local current_state = self:get_state()
    if current_state == ThreadState.SHUTDOWN then
        return
    end
    
    self:set_state(ThreadState.SHUTTING_DOWN)
    
    -- Clear cache and resources
    self:clear_cache()
    self:clear_resources()
    
    -- Call implementation-specific disposal
    self:dispose_impl()
    
    self:set_state(ThreadState.SHUTDOWN)
end

function ThreadSafeManager:dispose_impl()
    -- Override in subclasses
end

-- Thread-safe singleton pattern
local ThreadSafeSingleton = {}
ThreadSafeSingleton.__index = ThreadSafeSingleton

ThreadSafeSingleton._instances = {}
ThreadSafeSingleton._instance_locks = {}
ThreadSafeSingleton._singleton_lock = thread.newlock()

function ThreadSafeSingleton.new(name)
    name = name or "ThreadSafeSingleton"
    
    ThreadSafeSingleton._singleton_lock:lock()
    if not ThreadSafeSingleton._instances[name] then
        ThreadSafeSingleton._instances[name] = ThreadSafeManager.new(name)
        ThreadSafeSingleton._instance_locks[name] = thread.newlock()
    end
    ThreadSafeSingleton._singleton_lock:unlock()
    
    return ThreadSafeSingleton._instances[name]
end

function ThreadSafeSingleton.get_instance(name)
    return ThreadSafeSingleton.new(name)
end

function ThreadSafeSingleton.reset_instance(name)
    ThreadSafeSingleton._singleton_lock:lock()
    ThreadSafeSingleton._instances[name] = nil
    ThreadSafeSingleton._instance_locks[name] = nil
    ThreadSafeSingleton._singleton_lock:unlock()
end

-- Thread-safe event emitter
local ThreadSafeEventEmitter = {}
ThreadSafeEventEmitter.__index = ThreadSafeEventEmitter

function ThreadSafeEventEmitter.new()
    local self = setmetatable({}, ThreadSafeEventEmitter)
    self.lock = thread.newlock()
    self.listeners = {}
    self.event_stats = {}
    return self
end

function ThreadSafeEventEmitter:on(event, callback)
    self.lock:lock()
    if not self.listeners[event] then
        self.listeners[event] = {}
    end
    table.insert(self.listeners[event], callback)
    self.lock:unlock()
end

function ThreadSafeEventEmitter:off(event, callback)
    self.lock:lock()
    if self.listeners[event] then
        for i, listener in ipairs(self.listeners[event]) do
            if listener == callback then
                table.remove(self.listeners[event], i)
                break
            end
        end
    end
    self.lock:unlock()
end

function ThreadSafeEventEmitter:emit(event, ...)
    local listeners = {}
    self.lock:lock()
    if self.listeners[event] then
        listeners = self.listeners[event]
        self.event_stats[event] = (self.event_stats[event] or 0) + 1
    end
    self.lock:unlock()
    
    -- Call listeners outside of lock to prevent deadlocks
    for _, listener in ipairs(listeners) do
        local ok, err = pcall(listener, ...)
        if not ok then
            print("Error in event listener: " .. tostring(err))
        end
    end
end

function ThreadSafeEventEmitter:clear_listeners(event)
    self.lock:lock()
    if event then
        self.listeners[event] = nil
        self.event_stats[event] = nil
    else
        self.listeners = {}
        self.event_stats = {}
    end
    self.lock:unlock()
end

function ThreadSafeEventEmitter:get_event_stats()
    self.lock:lock()
    local stats = {}
    for event, count in pairs(self.event_stats) do
        stats[event] = count
    end
    self.lock:unlock()
    return stats
end

function ThreadSafeEventEmitter:get_listener_count(event)
    self.lock:lock()
    local count = self.listeners[event] and #self.listeners[event] or 0
    self.lock:unlock()
    return count
end

-- Export the module
return {
    ThreadSafeManager = ThreadSafeManager,
    ThreadSafeSingleton = ThreadSafeSingleton,
    ThreadSafeEventEmitter = ThreadSafeEventEmitter,
    ThreadSafeStats = ThreadSafeStats,
    ThreadState = ThreadState
} 

File: bindings/wasm/src/thread_safe_manager.ts
/**
 * Thread-Safe Manager for WASM Bindings
 * Provides comprehensive thread safety using Web Workers and SharedArrayBuffer
 */

// Thread state enumeration
export enum ThreadState {
    UNINITIALIZED = "uninitialized",
    INITIALIZING = "initializing",
    INITIALIZED = "initialized",
    SHUTTING_DOWN = "shutting_down",
    SHUTDOWN = "shutdown",
    ERROR = "error"
}

// Thread safety statistics
export interface ThreadSafeStats {
    totalOperations: number;
    concurrentOperations: number;
    maxConcurrentOperations: number;
    averageOperationTime: number;
    totalOperationTime: number;
    lockContentionCount: number;
    lastOperationTime: number;
}

export class ThreadSafeStatsImpl implements ThreadSafeStats {
    totalOperations = 0;
    concurrentOperations = 0;
    maxConcurrentOperations = 0;
    averageOperationTime = 0.0;
    totalOperationTime = 0.0;
    lockContentionCount = 0;
    lastOperationTime = 0.0;
}

// Event handler type
export type EventHandler = (data: any) => void;

// Lock implementation using SharedArrayBuffer
class SharedLock {
    private buffer: SharedArrayBuffer;
    private lockArray: Int32Array;
    private lockId: number;

    constructor() {
        this.buffer = new SharedArrayBuffer(4);
        this.lockArray = new Int32Array(this.buffer);
        this.lockId = Math.floor(Math.random() * 1000000);
    }

    async acquire(): Promise<void> {
        while (true) {
            const oldValue = Atomics.compareExchange(this.lockArray, 0, 0, this.lockId);
            if (oldValue === 0) {
                return;
            }
            await new Promise(resolve => setTimeout(resolve, 1));
        }
    }

    release(): void {
        Atomics.compareExchange(this.lockArray, 0, this.lockId, 0);
    }

    getBuffer(): SharedArrayBuffer {
        return this.buffer;
    }
}

// Thread-safe manager base class for WASM bindings
export class ThreadSafeManager {
    private name: string;
    private state: ThreadState = ThreadState.UNINITIALIZED;
    private initialized = false;
    private disposed = false;
    private lastError = "";
    private errorCount = 0;

    // Thread safety primitives
    private stateLock = new SharedLock();
    private errorLock = new SharedLock();
    private statsLock = new SharedLock();
    private operationLock = new SharedLock();
    private eventLock = new SharedLock();
    private resourceLock = new SharedLock();
    private configLock = new SharedLock();
    private cacheLock = new SharedLock();

    // Statistics
    private stats = new ThreadSafeStatsImpl();
    private operationStartTimes = new Map<number, number>();
    private activeOperations = 0;
    private operationCounter = 0;

    // Event handling
    private eventHandlers = new Map<string, EventHandler[]>();
    private eventStats = new Map<string, number>();

    // Resource management
    private resources = new Map<string, any>();

    // Configuration
    private config = new Map<string, any>();

    // Cache
    private cache = new Map<string, any>();
    private cacheTimestamps = new Map<string, number>();

    // Performance monitoring
    private performanceEnabled = true;
    private startTime = performance.now();

    // Worker management
    private workers: Worker[] = [];
    private workerLock = new SharedLock();

    constructor(name = "ThreadSafeManager") {
        this.name = name;
    }

    async ensureInitialized(): Promise<boolean> {
        await this.stateLock.acquire();
        try {
            if (this.state === ThreadState.UNINITIALIZED) {
                this.setError("Manager not initialized");
                return false;
            }
            if (this.state === ThreadState.SHUTDOWN) {
                this.setError("Manager has been disposed");
                return false;
            }
            if (this.state === ThreadState.ERROR) {
                this.setError("Manager is in error state");
                return false;
            }
            return true;
        } finally {
            this.stateLock.release();
        }
    }

    async ensureNotDisposed(): Promise<boolean> {
        await this.stateLock.acquire();
        try {
            if (this.state === ThreadState.SHUTDOWN) {
                this.setError("Manager has been disposed");
                return false;
            }
            return true;
        } finally {
            this.stateLock.release();
        }
    }

    async setState(state: ThreadState): Promise<void> {
        await this.stateLock.acquire();
        try {
            this.state = state;
        } finally {
            this.stateLock.release();
        }
    }

    async getState(): Promise<ThreadState> {
        await this.stateLock.acquire();
        try {
            return this.state;
        } finally {
            this.stateLock.release();
        }
    }

    async waitForState(targetState: ThreadState, timeout = 5000): Promise<boolean> {
        const startTime = performance.now();
        while (performance.now() - startTime < timeout) {
            if (await this.getState() === targetState) {
                return true;
            }
            await new Promise(resolve => setTimeout(resolve, 10));
        }
        return false;
    }

    async beginOperation(): Promise<number> {
        if (!this.performanceEnabled) {
            return 0;
        }

        await this.statsLock.acquire();
        try {
            const operationId = this.operationCounter++;
            this.activeOperations++;
            this.operationStartTimes.set(operationId, performance.now());
            this.stats.concurrentOperations = this.activeOperations;
            this.stats.maxConcurrentOperations = Math.max(
                this.stats.maxConcurrentOperations,
                this.activeOperations
            );
            return operationId;
        } finally {
            this.statsLock.release();
        }
    }

    async endOperation(operationId: number): Promise<void> {
        if (!this.performanceEnabled || operationId === 0) {
            return;
        }

        await this.statsLock.acquire();
        try {
            const startTime = this.operationStartTimes.get(operationId);
            if (startTime) {
                const operationTime = performance.now() - startTime;
                this.operationStartTimes.delete(operationId);

                this.activeOperations--;
                this.stats.totalOperations++;
                this.stats.totalOperationTime += operationTime;
                this.stats.averageOperationTime = this.stats.totalOperationTime / this.stats.totalOperations;
                this.stats.lastOperationTime = performance.now() - this.startTime;
                this.stats.concurrentOperations = this.activeOperations;
            }
        } finally {
            this.statsLock.release();
        }
    }

    async setError(error: string): Promise<void> {
        await this.errorLock.acquire();
        try {
            this.lastError = error;
            this.errorCount++;
            await this.setState(ThreadState.ERROR);
        } finally {
            this.errorLock.release();
        }
    }

    async clearError(): Promise<void> {
        await this.errorLock.acquire();
        try {
            this.lastError = "";
            if (this.state === ThreadState.ERROR) {
                await this.setState(ThreadState.INITIALIZED);
            }
        } finally {
            this.errorLock.release();
        }
    }

    async getLastError(): Promise<string> {
        await this.errorLock.acquire();
        try {
            return this.lastError;
        } finally {
            this.errorLock.release();
        }
    }

    async getErrorCount(): Promise<number> {
        await this.errorLock.acquire();
        try {
            return this.errorCount;
        } finally {
            this.errorLock.release();
        }
    }

    async isInitialized(): Promise<boolean> {
        return await this.getState() === ThreadState.INITIALIZED;
    }

    async isDisposed(): Promise<boolean> {
        return await this.getState() === ThreadState.SHUTDOWN;
    }

    async getStats(): Promise<ThreadSafeStats> {
        await this.statsLock.acquire();
        try {
            return {
                totalOperations: this.stats.totalOperations,
                concurrentOperations: this.stats.concurrentOperations,
                maxConcurrentOperations: this.stats.maxConcurrentOperations,
                averageOperationTime: this.stats.averageOperationTime,
                totalOperationTime: this.stats.totalOperationTime,
                lockContentionCount: this.stats.lockContentionCount,
                lastOperationTime: this.stats.lastOperationTime
            };
        } finally {
            this.statsLock.release();
        }
    }

    async resetStats(): Promise<void> {
        await this.statsLock.acquire();
        try {
            this.stats = new ThreadSafeStatsImpl();
            this.operationStartTimes.clear();
            this.activeOperations = 0;
            this.operationCounter = 0;
        } finally {
            this.statsLock.release();
        }
    }

    enablePerformanceMonitoring(): void {
        this.performanceEnabled = true;
    }

    disablePerformanceMonitoring(): void {
        this.performanceEnabled = false;
    }

    getUptime(): number {
        return performance.now() - this.startTime;
    }

    async setConfig(key: string, value: any): Promise<void> {
        await this.configLock.acquire();
        try {
            this.config.set(key, value);
        } finally {
            this.configLock.release();
        }
    }

    async getConfig(key: string, defaultValue?: any): Promise<any> {
        await this.configLock.acquire();
        try {
            return this.config.get(key) ?? defaultValue;
        } finally {
            this.configLock.release();
        }
    }

    async setCache(key: string, value: any, ttl = 300000): Promise<void> {
        await this.cacheLock.acquire();
        try {
            this.cache.set(key, value);
            this.cacheTimestamps.set(key, performance.now() + ttl);
        } finally {
            this.cacheLock.release();
        }
    }

    async getCache(key: string, defaultValue?: any): Promise<any> {
        await this.cacheLock.acquire();
        try {
            if (this.cache.has(key)) {
                const timestamp = this.cacheTimestamps.get(key);
                if (timestamp && performance.now() < timestamp) {
                    return this.cache.get(key);
                } else {
                    // Expired, remove from cache
                    this.cache.delete(key);
                    this.cacheTimestamps.delete(key);
                }
            }
            return defaultValue;
        } finally {
            this.cacheLock.release();
        }
    }

    async clearCache(): Promise<void> {
        await this.cacheLock.acquire();
        try {
            this.cache.clear();
            this.cacheTimestamps.clear();
        } finally {
            this.cacheLock.release();
        }
    }

    async addEventHandler(event: string, handler: EventHandler): Promise<void> {
        await this.eventLock.acquire();
        try {
            if (!this.eventHandlers.has(event)) {
                this.eventHandlers.set(event, []);
            }
            this.eventHandlers.get(event)!.push(handler);
        } finally {
            this.eventLock.release();
        }
    }

    async removeEventHandler(event: string, handlerIndex: number): Promise<void> {
        await this.eventLock.acquire();
        try {
            const handlers = this.eventHandlers.get(event);
            if (handlers && handlers[handlerIndex]) {
                handlers.splice(handlerIndex, 1);
            }
        } finally {
            this.eventLock.release();
        }
    }

    async emitEvent(event: string, data: any): Promise<void> {
        let handlers: EventHandler[] = [];
        await this.eventLock.acquire();
        try {
            if (this.eventHandlers.has(event)) {
                handlers = [...this.eventHandlers.get(event)!];
                this.eventStats.set(event, (this.eventStats.get(event) || 0) + 1);
            }
        } finally {
            this.eventLock.release();
        }

        // Call handlers outside of lock to prevent deadlocks
        for (const handler of handlers) {
            try {
                handler(data);
            } catch (error) {
                await this.setError(`Error in event handler: ${error}`);
            }
        }
    }

    async addResource(name: string, resource: any): Promise<void> {
        await this.resourceLock.acquire();
        try {
            this.resources.set(name, resource);
        } finally {
            this.resourceLock.release();
        }
    }

    async getResource<T>(name: string): Promise<T | undefined> {
        await this.resourceLock.acquire();
        try {
            return this.resources.get(name);
        } finally {
            this.resourceLock.release();
        }
    }

    async removeResource(name: string): Promise<void> {
        await this.resourceLock.acquire();
        try {
            this.resources.delete(name);
        } finally {
            this.resourceLock.release();
        }
    }

    async clearResources(): Promise<void> {
        await this.resourceLock.acquire();
        try {
            this.resources.clear();
        } finally {
            this.resourceLock.release();
        }
    }

    // Worker management for true parallelism
    async createWorker(script: string): Promise<Worker> {
        await this.workerLock.acquire();
        try {
            const worker = new Worker(script, { type: 'module' });
            this.workers.push(worker);
            return worker;
        } finally {
            this.workerLock.release();
        }
    }

    async terminateWorker(worker: Worker): Promise<void> {
        await this.workerLock.acquire();
        try {
            worker.terminate();
            const index = this.workers.indexOf(worker);
            if (index > -1) {
                this.workers.splice(index, 1);
            }
        } finally {
            this.workerLock.release();
        }
    }

    async terminateAllWorkers(): Promise<void> {
        await this.workerLock.acquire();
        try {
            for (const worker of this.workers) {
                worker.terminate();
            }
            this.workers = [];
        } finally {
            this.workerLock.release();
        }
    }

    async dispose(): Promise<void> {
        const currentState = await this.getState();
        if (currentState === ThreadState.SHUTDOWN) {
            return;
        }

        await this.setState(ThreadState.SHUTTING_DOWN);

        // Clear cache and resources
        await this.clearCache();
        await this.clearResources();
        await this.terminateAllWorkers();

        // Call implementation-specific disposal
        await this.disposeImpl();

        await this.setState(ThreadState.SHUTDOWN);
    }

    protected async disposeImpl(): Promise<void> {
        // Override in subclasses
    }
}

// Thread-safe singleton pattern
export class ThreadSafeSingleton<T> {
    private static instances = new Map<string, any>();
    private static instanceLocks = new Map<string, SharedLock>();
    private static singletonLock = new SharedLock();

    static async getOrCreate<T>(
        name: string,
        creator: () => Promise<T>
    ): Promise<T> {
        await ThreadSafeSingleton.singletonLock.acquire();
        try {
            if (!ThreadSafeSingleton.instances.has(name)) {
                const instance = await creator();
                ThreadSafeSingleton.instances.set(name, instance);
                ThreadSafeSingleton.instanceLocks.set(name, new SharedLock());
            }
            return ThreadSafeSingleton.instances.get(name);
        } finally {
            ThreadSafeSingleton.singletonLock.release();
        }
    }

    static async get<T>(name: string): Promise<T | undefined> {
        await ThreadSafeSingleton.singletonLock.acquire();
        try {
            return ThreadSafeSingleton.instances.get(name);
        } finally {
            ThreadSafeSingleton.singletonLock.release();
        }
    }

    static async resetInstance(name: string): Promise<void> {
        await ThreadSafeSingleton.singletonLock.acquire();
        try {
            ThreadSafeSingleton.instances.delete(name);
            ThreadSafeSingleton.instanceLocks.delete(name);
        } finally {
            ThreadSafeSingleton.singletonLock.release();
        }
    }
}

// Thread-safe event emitter
export class ThreadSafeEventEmitter {
    private lock = new SharedLock();
    private listeners = new Map<string, EventHandler[]>();
    private eventStats = new Map<string, number>();

    async on(event: string, callback: EventHandler): Promise<void> {
        await this.lock.acquire();
        try {
            if (!this.listeners.has(event)) {
                this.listeners.set(event, []);
            }
            this.listeners.get(event)!.push(callback);
        } finally {
            this.lock.release();
        }
    }

    async off(event: string, callback: EventHandler): Promise<void> {
        await this.lock.acquire();
        try {
            const handlers = this.listeners.get(event);
            if (handlers) {
                const index = handlers.indexOf(callback);
                if (index > -1) {
                    handlers.splice(index, 1);
                }
            }
        } finally {
            this.lock.release();
        }
    }

    async emit(event: string, data: any): Promise<void> {
        let handlers: EventHandler[] = [];
        await this.lock.acquire();
        try {
            if (this.listeners.has(event)) {
                handlers = [...this.listeners.get(event)!];
                this.eventStats.set(event, (this.eventStats.get(event) || 0) + 1);
            }
        } finally {
            this.lock.release();
        }

        // Call handlers outside of lock to prevent deadlocks
        for (const handler of handlers) {
            try {
                handler(data);
            } catch (error) {
                console.error("Error in event listener:", error);
            }
        }
    }

    async clearListeners(event?: string): Promise<void> {
        await this.lock.acquire();
        try {
            if (event) {
                this.listeners.delete(event);
                this.eventStats.delete(event);
            } else {
                this.listeners.clear();
                this.eventStats.clear();
            }
        } finally {
            this.lock.release();
        }
    }

    async getEventStats(): Promise<Map<string, number>> {
        await this.lock.acquire();
        try {
            return new Map(this.eventStats);
        } finally {
            this.lock.release();
        }
    }

    async getListenerCount(event: string): Promise<number> {
        await this.lock.acquire();
        try {
            return this.listeners.get(event)?.length || 0;
        } finally {
            this.lock.release();
        }
    }
}

// Utility for creating thread-safe promises
export class ThreadSafePromise<T> {
    private promise: Promise<T>;
    private resolve!: (value: T) => void;
    private reject!: (reason: any) => void;
    private lock = new SharedLock();

    constructor() {
        this.promise = new Promise<T>((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
    }

    async resolve(value: T): Promise<void> {
        await this.lock.acquire();
        try {
            this.resolve(value);
        } finally {
            this.lock.release();
        }
    }

    async reject(reason: any): Promise<void> {
        await this.lock.acquire();
        try {
            this.reject(reason);
        } finally {
            this.lock.release();
        }
    }

    getPromise(): Promise<T> {
        return this.promise;
    }
} 